README
DISCLAIMER

This is the UNSTABLE, possibly untested, DEVELOPER RELEASE of
String-Sortable to GitHub.

This module has not yet been released to CPAN.

END OF DISCLAIMER

-8<-


String::Sortable
================

SYNOPSIS

    use String::Sortable;

    my $sortable = String::Sortable->new( '¬The Title of ¬the Thing' );

    printf "Display: %s\n", $sortable->display;  # The Title of the Thing
    printf "Sortify: %s\n", $sortable->sortify;  # title of thing


INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


DOCUMENTATION

After installing, you can find documentation for this module with
the perldoc command:

    perldoc String::Sortable


COPYRIGHT AND LICENSE

Copyright (c) 2010-2011 Henning Manske. All rights reserved.

This module is free software. You can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/

This module is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


Changes
Revision history for Perl extension String-Sortable

0.00_03 May 27, 2011

        - Minor cosmetic changes

0.00_02 Dec 17, 2010

        - Major refactoring:
            * User-defined control characters
            * Remove auto-detection of markup
            * Setup defaults via import
            * Export pre-configured functions on demand
        - Switch GitHub URLs to https
        - Explicitly require Test::More 0.62 in tests

0.00_01 Sep 22, 2010

        - Initial commit and release to GitHub


xt/author/Test/EOL.pm
package Test::EOL;
BEGIN {
  $Test::EOL::AUTHORITY = 'cpan:FLORA';
}
BEGIN {
  $Test::EOL::VERSION = '0.9';
}
# ABSTRACT: Check the correct line endings in your project

use strict;
use warnings;

use Test::Builder;
use File::Spec;
use FindBin qw($Bin);
use File::Find;

use vars qw( $PERL $UNTAINT_PATTERN $PERL_PATTERN);

$PERL    = $^X || 'perl';
$UNTAINT_PATTERN  = qr|^([-+@\w./:\\]+)$|;
$PERL_PATTERN     = qr/^#!.*perl/;

my %file_find_arg = ($] <= 5.006) ? () : (
    untaint => 1,
    untaint_pattern => $UNTAINT_PATTERN,
    untaint_skip => 1,
);

my $Test  = Test::Builder->new;
my $updir = File::Spec->updir();

sub import {
    my $self   = shift;
    my $caller = caller;
    {
        no strict 'refs';
        *{$caller.'::eol_unix_ok'} = \&eol_unix_ok;
        *{$caller.'::all_perl_files_ok'} = \&all_perl_files_ok;
    }
    $Test->exported_to($caller);
    $Test->plan(@_);
}

sub _all_perl_files {
    my @all_files = _all_files(@_);
    return grep { _is_perl_module($_) || _is_perl_script($_) } @all_files;
}

sub _all_files {
    my @base_dirs = @_ ? @_ : File::Spec->catdir($Bin, $updir);
    my @found;
    my $want_sub = sub {
        return if ($File::Find::dir =~ m![\\/]?CVS[\\/]|[\\/]?.svn[\\/]!); # Filter out cvs or subversion dirs/
        return if ($File::Find::dir =~ m![\\/]?blib[\\/]libdoc$!); # Filter out pod doc in dist
        return if ($File::Find::dir =~ m![\\/]?blib[\\/]man\d$!); # Filter out pod doc in dist
        return if ($File::Find::name =~ m!Build$!i); # Filter out autogenerated Build script
        return unless (-f $File::Find::name && -r _);
        push @found, File::Spec->no_upwards( $File::Find::name );
    };
    my $find_arg = {
        %file_find_arg,
        wanted   => $want_sub,
        no_chdir => 1,
    };
    find( $find_arg, @base_dirs);
    return @found;
}

# Formats various human invisible symbols
# to similar visible ones.
# Perhaps ^M or something like that
# would be more appropriate?

sub _show_whitespace {
    my $string = shift;
    $string =~ s/\r/[\\r]/g;
    $string =~ s/\t/[\\t]/g;
    $string =~ s/ /[\\s]/g;
    return $string;
}

# Format a line record for diagnostics.

sub _debug_line {
    my ( $options, $line ) = @_;
    $line->[2] =~ s/\n\z//g;
    return "line $line->[1] : $line->[0] " . (
      $options->{show_lines} ? qq{: } . _show_whitespace( $line->[2] )  : q{}
    );
}

sub eol_unix_ok {
    my $file = shift;
    my $test_txt;
    $test_txt   = shift if !ref $_[0];
    $test_txt ||= "No windows line endings in '$file'";
    my $options = shift if ref $_[0] eq 'HASH';
    $options ||= {
        trailing_whitespace => 0,
        all_reasons => 0,
    };
    $file = _module_to_path($file);

    open my $fh, $file or do { $Test->ok(0, $test_txt); $Test->diag("Could not open $file: $!"); return; };
    # Windows-- , default is :crlf, which hides \r\n  -_-

#  patch applied from
#  RT #59877: Test::EOL should not assume UTF-8 encoding on tested files

#   binmode( $fh, ':raw:utf8' );
    binmode( $fh, ':raw' );
    my $line = 0;
    my @fails;
    while (<$fh>) {
        $line++;
        if ( !$options->{trailing_whitespace} && /(\r+)$/ ) {
          my $match = $1;
          push @fails, [ _show_whitespace( $match ) , $line , $_ ];
        }
        if (  $options->{trailing_whitespace} && /([ \t]*\r+|[ \t]+)$/ ) {
          my $match = $1;
          push @fails, [ _show_whitespace($match), $line , $_ ];
        }
        # Minor short-circuit for people who don't need the whole file scanned
        # once there's an err.
        last if( @fails > 0 && !$options->{all_reasons} );
    }
    if( @fails ){
       $Test->ok( 0, $test_txt . " on "  . _debug_line({ show_lines => 0 } , $fails[0]  )  );
       if ( $options->{all_reasons} || 1 ){
          $Test->diag( "  Problem Lines: ");
          for ( @fails ){
            $Test->diag(_debug_line({ show_lines => 1 } , $_ ) );
          }
       }
       return 0;
    }
    $Test->ok(1, $test_txt);
    return 1;
}
sub all_perl_files_ok {
    my $options = shift if ref $_[0] eq 'HASH';
    my @files = _all_perl_files( @_ );
    _make_plan();
    foreach my $file ( @files ) {
      eol_unix_ok($file, $options);
    }
}

sub _is_perl_module {
    $_[0] =~ /\.pm$/i || $_[0] =~ /::/;
}

sub _is_perl_script {
    my $file = shift;
    return 1 if $file =~ /\.pl$/i;
    return 1 if $file =~ /\.t$/;
    open (my $fh, $file) or return;
    my $first = <$fh>;
    return 1 if defined $first && ($first =~ $PERL_PATTERN);
    return;
}

sub _module_to_path {
    my $file = shift;
    return $file unless ($file =~ /::/);
    my @parts = split /::/, $file;
    my $module = File::Spec->catfile(@parts) . '.pm';
    foreach my $dir (@INC) {
        my $candidate = File::Spec->catfile($dir, $module);
        next unless (-e $candidate && -f _ && -r _);
        return $candidate;
    }
    return $file;
}

sub _make_plan {
    unless ($Test->has_plan) {
        $Test->plan( 'no_plan' );
    }
    $Test->expected_tests;
}

sub _untaint {
    my @untainted = map { ($_ =~ $UNTAINT_PATTERN) } @_;
    return wantarray ? @untainted : $untainted[0];
}

1;


__END__
=pod

=encoding utf-8

=head1 NAME

Test::EOL - Check the correct line endings in your project

=head1 SYNOPSIS

C<Test::EOL> lets you check the presence of windows line endings in your
perl code. It
report its results in standard C<Test::Simple> fashion:

  use Test::EOL tests => 1;
  eol_unix_ok( 'lib/Module.pm', 'Module is ^M free');

and to add checks for trailing whitespace:

  use Test::EOL tests => 1;
  eol_unix_ok( 'lib/Module.pm', 'Module is ^M and trailing whitespace free', { trailing_whitespace => 1 });

Module authors can include the following in a t/eol.t and have C<Test::EOL>
automatically find and check all perl files in a module distribution:

  use Test::EOL;
  all_perl_files_ok();

or

  use Test::EOL;
  all_perl_files_ok( @mydirs );

and if authors would like to check for trailing whitespace:

  use Test::EOL;
  all_perl_files_ok({ trailing_whitespace => 1 });

or

  use Test::EOL;
  all_perl_files_ok({ trailing_whitespace => 1 }, @mydirs );

=head1 DESCRIPTION

This module scans your project/distribution for any perl files (scripts,
modules, etc) for the presence of windows line endings.

=head1 FUNCTIONS

=head2 all_perl_files_ok( [ \%options ], [ @directories ] )

Applies C<eol_unix_ok()> to all perl files found in C<@directories> (and sub
directories). If no <@directories> is given, the starting point is one level
above the current running script, that should cover all the files of a typical
CPAN distribution. A perl file is *.pl or *.pm or *.t or a file starting
with C<#!...perl>

If the test plan is defined:

  use Test::EOL tests => 3;
  all_perl_files_ok();

the total number of files tested must be specified.

=head2 eol_unix_ok( $file [, $text] [, \%options ]  )

Run a unix EOL check on C<$file>. For a module, the path (lib/My/Module.pm) or the
name (My::Module) can be both used.

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 ACKNOWLEDGEMENTS

Shamelessly ripped off from L<Test::NoTabs>.

=head1 SEE ALSO

L<Test::More>, L<Test::Pod>. L<Test::Distribution>, L<Test:NoWarnings>,
L<Test::NoTabs>, L<Module::Install::AuthorTests>.

=head1 AUTHORS

=over 4

=item *

Tomas Doran <bobtfish@bobtfish.net>

=item *

Arthur Axel 'fREW' Schmidt <frioux@gmail.com>

=item *

Kent Fredric <kentfredric@gmail.com>

=item *

Florian Ragwitz <rafl@debian.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Tomas Doran.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/String/Sortable.pm
package String::Sortable;

use 5.006;

use strict;
use warnings FATAL => 'all';

our $VERSION = '0.00_03';

use Carp ();

use constant NONSORT_CC   => chr(172); # ¬
use constant SORT_CC      => '@';
use constant EMPTY_STRING => q{};

use overload fallback => 1,
  '""'  => 'display',
  'cmp' => 'compare',
  'eq'  => 'equals',
  'ne'  => sub { ! shift->equals(shift) };



{
  my $_setup = { nonsort => NONSORT_CC };

  sub setup {
    my $class = shift;
    if (@_) {
      $_setup = _prepare(shift);
    }
    return $_setup;
  }
}



sub _prepare {
  my $in = shift;
  my %out;

  if ($in) {
    if (ref $in eq 'HASH') {
      %out = %$in;
      if (defined $out{nonsort}) {
        $out{sort} = undef;
      }
    }
    elsif ($in eq 'sort') {
      $out{sort} = SORT_CC;
    }
    elsif ($in eq 'nonsort') {
      $out{sort} = undef;
      $out{nonsort} = NONSORT_CC;
    }
    else {
      Carp::croak "Unknown parameter: $in";
    }
  }

  my $setup = setup;

  for (keys %$setup) {
    $out{$_} = $setup->{$_} unless exists $out{$_};
  }

  if (exists $out{nonsort}) {
    if (defined $out{sort}) {
      delete $out{nonsort};
    }
    elsif (!defined $out{nonsort}) {
      if (exists $out{sort}) {
        delete $out{sort};
        $out{nonsort} = NONSORT_CC;
      }
      else {
        delete $out{nonsort};
        $out{sort} = SORT_CC;
      }
    }
    elsif (exists $out{sort}) {
      delete $out{sort};
    }
  }
  elsif (!defined $out{sort}) {
    delete $out{sort};
    $out{nonsort} = NONSORT_CC;
  }

  return \%out;
}



sub import {
  my $class = shift;

  while (@_) {
    my $name   = shift;
    my $params = shift;

    next unless defined $name;

    if ($name eq '-setup') {
      $class->setup($params);
    }
    else {
      my $caller = caller;
      $params = _prepare($params);

      no strict 'refs';
      *{$caller . '::' . $name} = sub { map { $class->new($_, %$params) } @_ };
    }
  }
}



#
#  constructor
#

sub new {
  my $class = shift;
  unshift @_, 'raw' if @_ % 2;
  bless { @_ }, $class;
}



#
#  public methods
#

sub display {
  my $self = shift;
  return exists $self->_parse->{display} ? $self->{display} : $self->{raw};
}



sub sortify {
  my $self = shift;
  return exists $self->_parse->{sortified} ? $self->{sortified} :
    exists $self->{display} ? lc $self->{display} : lc $self->{raw};
}



sub compare {
  my ($self, $other, $reversed) = @_;
  my $result = $self->sortify cmp
    ( eval { $other->isa(__PACKAGE__) } ? $other->sortify : $other );
  return $reversed ? $result * -1 : $result;
}



sub equals {
  my ($self, $other) = @_;
  return $self->display eq
    ( eval { $other->isa(__PACKAGE__) } ? $other->display : $other );
}



#
#  private methods
#

{
  my %_cache;

  sub _parse {
    my $self = shift;
    return $self if $self->{_parsed}++;

    if (! defined(my $raw = $self->{raw}) ) {
      $self->{display} = EMPTY_STRING;
    }
    else {
      my ($cc, $sort);

      if (defined($cc = $self->{sort})) {
        $sort = 1;
      }
      else {
        unless (defined($cc = $self->{nonsort})) {
          my $setup = $self->setup;
          if (defined($cc = $setup->{sort})) {
            $sort = 1;
          }
          else {
            $cc = $setup->{nonsort};
          }
        }
      }

      if ((my $index = index $raw, $cc) > -1) {
        if ($sort) {
          if ($index == 0) {
            $self->{display} = substr $raw, 1;
          }
          else {
            $self->{display} = substr($raw, 0, $index++) . substr $raw, $index;
            $self->{sortified} = lc substr $raw, $index;
          }
        }
        else {
          my $pattern = $_cache{$cc} ||= do {
            $cc = quotemeta $cc;
            qr{
              $cc
              ( [[:alnum:]]*    # any alphanumerical character
                [ .]?           # optional space or full stop
                (?:'(?=$cc))?   # optional apostrophe, if followed by the control character
              )
              (?: (?<=')$cc )?  # control character, if preceeded by apostrophe
              ( [^$cc]* )       # anything except the control character
            }x
          };
          my $display = my $sortified = EMPTY_STRING;
          while ($raw =~ s/$pattern//) {
            $display .= $1 . $2;
            $sortified .= $2;
          }
          $self->{display} = $display;
          $self->{sortified} = lc $sortified;
        }
      }
    }

    return $self
  }

} # end of scope of %_cache



$VERSION = eval $VERSION;

__END__

=head1 NAME

String::Sortable - Perl class for strings containing sorting information

=head1 SYNOPSIS

  use String::Sortable;

  my $sortable = String::Sortable->new( '¬The Title of ¬the Thing' );

  printf "Display: %s\n", $sortable->display;  # The Title of the Thing
  printf "Sortify: %s\n", $sortable->sortify;  # title of thing

=head1 DESCRIPTION

TODO.

=head1 CONSTRUCTOR

=over 4

=item $sortable = String::Sortable->new(%options);

Creates a new C<String::Sortable> object and returns it.

=item $sortable = String::Sortable->new($raw_string);

=item $sortable = String::Sortable->new($raw_string, %options);

=back

=head1 METHODS

=over 4

=item $config    = String::Sortable->setup($config);

=item $display   = $sortable->display;

=item $sortified = $sortable->sortify;

=item $result    = $sortable->compare($thing);

=item $boolean   = $sortable->equals($thing);

=back

=head1 TODO

=over 4

=item * Documentation

=item * C<%options>

=back

=head1 AUTHOR

Henning Manske <hma@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2010-2011 Henning Manske. All rights reserved.

This module is free software. You can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/>.

This module is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut


Makefile.PL
use 5.006;

use strict;
use warnings;

use ExtUtils::MakeMaker 6.17;

my $eumm = eval ExtUtils::MakeMaker->VERSION;

my $NAME = 'String::Sortable';

(my $dist = $NAME) =~ s|::|-|g;
(my $path = $NAME) =~ s|::|/|g;

my $META = {
  no_index => {
    directory  => 'xt',
  },
  resources => {
   #homepage   => "http://search.cpan.org/dist/$dist/",
    bugtracker => "https://github.com/hma/$dist/issues",
    repository => "https://github.com/hma/$dist",
    license    => 'http://dev.perl.org/licenses/',
  },
};

WriteMakefile(
  NAME          => $NAME,
  AUTHOR        => 'Henning Manske <hma@cpan.org>',
  VERSION_FROM  => "lib/$path.pm",
  ABSTRACT_FROM => "lib/$path.pm",

  ( $eumm >= 6.3001 ? ( LICENSE            => 'perl' ) : () ),
  ( $eumm >= 6.46   ? ( META_MERGE         => $META  ) : () ),
  ( $eumm >= 6.4701 ? ( MIN_PERL_VERSION   => 5.006  ) : () ),
  ( $eumm >= 6.5103 ? ( CONFIGURE_REQUIRES => { 'ExtUtils::MakeMaker' => '6.17' } ) : () ),
  ( $eumm >= 6.5503 ?  'BUILD_REQUIRES' : PREREQ_PM => {
    'Test::Deep' => '0',
    'Test::More' => '0.62',
  } ),

  dist  => { COMPRESS => 'gzip -9f' },
  clean => { FILES    => "$dist-* cover_db Debian_CPANTS.txt nytprof*" },
  test  => { TESTS    => 't/*.t xt/*.t' },
);
